src/libraries/System.IO.Pipelines/tests/PipeReaderWriterFacts.cs

```diff
@@ -229,6 +229,34 @@ void ThrowTestException()
             Assert.Single(Regex.Matches(invalidOperationException.StackTrace, "Pipe.GetReadResult"));
         }
 
+        [Fact]
+        public async Task WriteAsync_ThrowsIfReaderCompletedWithException()
+        {
+            void ThrowTestException()
+            {
+                try
+                {
+                    throw new InvalidOperationException("Reader exception");
+                }
+                catch (Exception e)
+                {
+                    _pipe.Reader.Complete(e);
+                }
+            }
+
+            ThrowTestException();
```

Comment:
This is a little fragile, in that it's relying on ThrowTestException not getting inlined in order for the test to succeed.  Today it won't be inlined, including because of the exception handling logic, but it'd be more reliable against future JIT improvements if it were attributed with MethodImplOptions.NoInlining (which with C# rules today would also require it to be a non-local method).